include/nuttx/list.h

```diff
@@ -79,192 +81,149 @@
+      __list->prev->next = __item; \
+      __list->prev       = __item; \
     } \
   while (0)
 
 #define list_delete(item) \
   do \
     { \
-      (item)->next->prev = (item)->prev; \
-      (item)->prev->next = (item)->next; \
-      (item)->prev       = (item)->next = NULL; \
+      FAR struct list_node *__item = (item); \
+      __item->next->prev = __item->prev; \
+      __item->prev->next = __item->next; \
+      __item->prev = __item->next = NULL; \
     } \
   while (0)
 
-#define list_remove_head_type(list, type, element) ( \
-{\
-  FAR struct list_node *__nod = list_remove_head(list); \
-  FAR type *__t; \
-\
-  if(__nod) \
-    { \
-      __t = container_of(__nod, type, element); \
-    } \
-  else \
-    {\
-      __t = (FAR type *)NULL; \
-    } \
-\
-  __t; \
-})
-
-#define list_remove_tail_type(list, type, element) ( \
-{\
-  FAR struct list_node *__nod = list_remove_tail(list); \
-  FAR type *__t; \
-\
-  if(__nod) \
-    { \
-      __t = container_of(__nod, type, element); \
-    } \
-  else\
-    { \
-      __t = (FAR type *)NULL; \
-    } \
-\
-  __t; \
-})
-
-#define list_peek_head_type(list, type, element) ( \
-{\
-  FAR struct list_node *__nod = list_peek_head(list); \
-  FAR type *__t; \
-\
-  if(__nod) \
-    { \
-      __t = container_of(__nod, type, element); \
-    } \
-  else \
-    { \
-      __t = (FAR type *)NULL; \
-    } \
-\
-  __t; \
-})
-
-#define list_peek_tail_type(list, type, element) ( \
-{ \
-  FAR struct list_node *__nod = list_peek_tail(list); \
-  FAR type *__t; \
-\
-  if(__nod) \
-    { \
-      __t = container_of(__nod, type, element); \
-    } \
-  else\
-    { \
-      __t = (FAR type *)NULL; \
-    } \
-\
-  __t; \
-})
-
-#define list_prev_type(list, item, type, element) ( \
-{ \
-  FAR struct list_node *__nod = list_prev(list, item); \
-  FAR type *__t; \
-\
-  if(__nod) \
-    { \
-      __t = container_of(__nod, type, element); \
-    } \
-  else \
-    { \
-      __t = (FAR type *)NULL; \
-    } \
-\
-  __t; \
-})
-
-#define list_prev_wrap_type(list, item, type, element) ( \
-{ \
-  FAR struct list_node *__nod = list_prev_wrap(list, item); \
-  FAR type *__t; \
-\
-  if(__nod) \
-    { \
-      __t = container_of(__nod, type, element); \
-    } \
-  else \
-    { \
-      __t = (FAR type *)NULL; \
-    } \
-\
-  __t; \
-})
-
-#define list_next_type(list, item, type, element) ( \
-{ \
-  FAR struct list_node *__nod = list_next(list, item); \
-  FAR type *__t; \
-\
-  if(__nod) \
-    { \
-      __t = container_of(__nod, type, element); \
-    } \
-  else \
-    { \
-      __t = (FAR type *)NULL; \
-    } \
-\
-  __t; \
-})
-
-#define list_next_wrap_type(list, item, type, element) ( \
-{\
-  FAR struct list_node *__nod = list_next_wrap(list, item); \
-  FAR type *__t; \
-\
-  if(__nod) \
-    { \
-      __t = container_of(__nod, type, element); \
-    } \
-  else \
-    { \
-      __t = (FAR type *)NULL; \
-    } \
-\
-  __t; \
-})
+#define list_remove_head_type(list, type, member) \
```

Comment:
This seems to be a GNU extension and is not supported by all compilers

Reply:
Yes, GNU extension can't be avoided here even we change macro to normal(no inline) function.